### ISSUE TITLE
[BUG] r27c NDK: incorrect nullability on fields in `group` and `passwd` structs

### ISSUE THREAD

**Issue by hyp** [android/ndk#2098]
### Description

The Android NDK includes nullability annotations on some `char *` fields and parameters. Unfortunately it looks like some of the nullability annotations on fields of the `group` and `passwd` structs from `grp.h` and `pwd.h` headers are incorrect.
I looked over the sources for Android's libc and it looks like the situation is mixed:

- `pw_name` doesn't seem to be nullable , so I think it should be `_Nonnull`  instead of being `_Nullable` in the `passwd` struct.
- `pw_dir` doesn't seem to be nullable, so I think it should be `_Nonnul` instead in the `passwd` struct.
- `gr_name` doesn't seem to be nullable, so I think it should be `_Nonnull` instead in the `group` struct.

This is problematic as Clang depends on these nullability annotations when the NDK headers are being imported into Swift, and these values are marked as Optional in Swift, which is different to all other POSIX platforms. This causes an issue when building swift-foundation for Android:
https://github.com/swiftlang/swift-foundation/pull/999

### Affected versions

r27

### Host OS

Windows

### Host OS version

Windows 11

### Affected ABIs

arm64-v8a

### Build system

ndk-build

### Other build system

Cmake

### minSdkVersion

28

**Comment by enh-google** [comment]
yeah, this isn't great but i'm unconvinced it's _wrong_... i've replied in more detail at https://github.com/swiftlang/swift-foundation/pull/999#discussion_r1821386117.

**Comment by enh-google** [comment]
(on the principle that me speaking up there might at least help you get your workaround in :-) that workaround seems like the least worst option to me atm, though i'm willing to be persuaded by more examples/other precedent.)

**Comment by DanAlbert** [comment]
@enh-google, there's no response to your comment on the other thread, but the change was merged. I take it you're in favor of closing this unless we see stronger evidence that bionic's wrong?

**Comment by DanAlbert** [comment]
> explicit initialization (= {}) defaults to 0/null...

I interpret `_Nonnull` fields to mean default initialization is invalid for those structs.

> of course you can combine that argument with my earlier the "but clang doesn't make much use of these annotations anyway" argument and get back to the "maybe we should just treat these as documentation, and say what's intended?". but i'm still not sure how useful that is, or to whom. (or, indeed, whether it's less useful than being honest.)

Hmm. On the one hand, because the guarantees from Clang are currently so weak, maybe it is better for Rust/Swift bindings to treat these properties as optional, even if no correctly initialized struct would have null values, just so it's possible to write code that defends against buggy C code. On the other hand, that makes the interface in those less-error-prone languages worse just so that they can defend against bugs that happened earlier in the program, and really that verification should happen earlier, before it leaves C.

If I'm understanding you correctly, bionic has annotated these `_Nullable` not because it's valid for an initialized struct to have null values, but because it's C, and you're perfectly welcome to shoot yourself in the foot. That's not a very good reason IMO. Clang may not offer much defense today, but we should be using the annotations however we would if they _were_ more thorough. We should not have to rewrite all of them when Clang eventually improves.

Also, like you say, documenting intent is important here, as long as intent means "always" and not "usually". If the only reason for the field to be `_Nullable` is because it might be temporarily null because you default initialized it, maybe that should just be `_Nonnull` and you shouldn't be using default initialization for that struct?

**Comment by enh-google** [comment]
> That's not a very good reason IMO. Clang may not offer much defense today, but we should be using the annotations however we would if they were more thorough. We should not have to rewrite all of them when Clang eventually improves.

i think this really depends on what (if anything) clang _will_ do. do we assume it's actually checking that you've had a [successful!] call to one of the functions that set up the "invariants"? and also checking that you haven't _undone_ that. and checking that -- when these things are pointers to shared data -- that you can't end up with a failed call on another thread leaving you with undefined state on your original thread? and -- since there's no `final` -- that code between you and us hasn't clobbered the field. (there's also the tricky job of ensuring that the implementation never puts a null in any of those fields in the first place. do we _know_ it doesn't? in theory clang could help check that, but in practice, i'm unconvinced it'll ever get clever enough, and it's certainly not doing much right now.)

i think there's a stronger argument that struct fields should be `_Null_unspecified` for now, and we re-evaluate if/when clang actually does something useful. or we do the
```
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wnullability-completeness"
...
#pragma clang diagnostic pop
```
hack around structs until/unless clang does something useful with this information. (that's actually what we've done for [i think] the majority of structs.)

**Comment by DanAlbert** [comment]
> that you can't end up with a failed call on another thread leaving you with undefined state on your original thread?

In general I'd say that data shouldn't be left in an undefined state on failure... but within bionic it's often the kernel making that choice so we don't have a choice. I think that argument alone is enough to convince me, at least for this case. The best things these annotations can do today is allow things like bindgen to generate more accurate interfaces, and given that we're not in control of the code that would need to guarantee non-nullness here, nullable is the more accurate description.
